### PR TITLE
Rds backups

### PR DESCRIPTION
**Pull request by DrizzlyOwl**
New Feature to add scheduled database backups that push .sql files over to an S3 bucket for each infrastructure.

Todo:
1. Finalise task definition and link it to an ECS Service so it can run
2. Determine if we want to host the container image available here: https://github.com/dxw/dalmatian-sql-backup on GitHub (and authenticate with a token) or on ECR. I think ECR would be best if we can add a workflow to the repo that pushes up to the ECR?
3. Determine if we need cloudwatch logs for this task? If so, they need connecting up to push the logs to the same place as the other ECS tasks

### CODE REVIEW

**Comment by DrizzlyOwl** [comment]
We can lose the changes to the s3 policies

File: locals.tf [hunk 1]
```diff
@@ -158,6 +158,39 @@ locals {
...
+      ${local.infrastructure_rds_backups_command_preamble} && \
+      PGPASSWORD="$DB_ROOT_PASSWORD"
+      psql -U "$DB_USER" -h "$DB_HOSTNAME" -t -c 'SELECT datname FROM pg_database WHERE NOT datistemplate' | \
+      while read DB_NAME; do
+        if [[ -n "$DB_NAME" && "$DB_NAME" != "rdsadmin" ]]; then
+          pg_dump --clean --if-exists \postgres://$DB_USER:$DB_ROOT_PASSWORD@$DB_HOSTNAME:5432/$DB_NAME > "/tmp/sqlbackups/$DB_NAME/$DATE_STRING-$DB_NAME.sql";
+        fi;
+      done && \
+      ${local.infrastructure_rds_backups_command_preamble}"
+    BASH
```
Comment (DrizzlyOwl) [inline locals.tf:191]: There's probably a nicer way to do this

File: rds-infrastructure-backups-cloudwatch.tf [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+resource "random_integer" "infrastructure_rds_daily_backups_hour" {
+  for_each = local.infrastructure_rds_backups_enabled
+
+  min = 0
+  max = 6
+  keepers = {
+    # Generate a new value if the db instance changes
+    listener_arn = aws_db_instance.infrastructure_rds[each.key].arn
+  }
+}
```
Comment (DrizzlyOwl) [inline rds-infrastructure-backups-cloudwatch.tf:11]: Using a random int to represent an hour between 00 and 06 means we can stagger the execution of the database backups for each instance so not all of our RDS instances get hammered at the same time

File: rds-infrastructure-backups-s3.tf [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+  bucket = aws_s3_bucket.infrastructure_rds_daily_backups[each.key].id
+
+  rule {
+    apply_server_side_encryption_by_default {
+      kms_master_key_id = local.infrastructure_kms_encryption ? aws_kms_key.infrastructure[0].arn : null
+      sse_algorithm     = local.infrastructure_kms_encryption ? "aws:kms" : "AES256"
+    }
+  }
+}
```
Comment (DrizzlyOwl) [inline rds-infrastructure-backups-s3.tf:84]: Worth testing whether `tfsec` is happy again now we have refactored the `rule {}` block

File: rds-infrastructure-backups-task.tf [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      secrets = jsonencode({
+        "DB_HOSTNAME"          = sensitive(aws_db_instance.infrastructure_rds[each.key].address)
+        "DB_ROOT_PASSWORD"     = sensitive(aws_secretsmanager_secret_version.infrastructure_rds_root_password[each.key].secret_string)
+        "DB_USER"              = "root"
+        "AWS_S3_BUCKET_TARGET" = aws_s3_bucket.infrastructure_rds_daily_backups[each.key].arn
+      })
+    }
+  )
+  execution_role_arn = aws_iam_role.infrastructure_rds_daily_backups_task_execution[each.key].arn
+  # task_role_arn            = aws_iam_role.infrastructure_ecs_cluster_service_task[each.key].arn
```
Comment (DrizzlyOwl) [inline rds-infrastructure-backups-task.tf:99]: Missing task role arn for the task definition and the cloudwatch event

**Comment by DrizzlyOwl** [comment]
Related PR https://github.com/dxw/dalmatian/pull/710

**Comment by shuldt** [comment]
Hi - is this still something we want to go in?

**Comment by DrizzlyOwl** [comment]
> Hi - is this still something we want to go in?

Hello, yes this is porting functionality over from v1 Dalmatian to v2 :)

**Comment by DrizzlyOwl** [comment]
Closing due to https://github.com/dxw/terraform-dxw-dalmatian-infrastructure/pull/136 happening before this was finished

**Pull request closed without merging**